### PR TITLE
[TikTok] Iterate through app versions

### LINKED ISSUE #2476: TikTok "aweme/detail" API endpoint returns an empty response; Falls back to "feed" endpoint
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.01.21**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

USA

### Description

"aweme/detail" API endpoint returns an empty response; Falls back to "feed" endpoint.

Probably entails repairing or removing the request to "aweme/detail" API endpoint.

Maybe this issue has some overlap with this pull request?:  https://github.com/yt-dlp/yt-dlp/pull/2449

I'm reporting it anyway, just in case.

I will try changing the _APP_VERSION (for one or both endpoints) and see if this resolves the issue. If it does, then I will post back here and let you know.

### Verbose log

```shell
root@vmi298562:/tmp/yt-dlp# python3.7 /usr/local/bin/yt-dlp -v --write-pages https://www.tiktok.com/@mimiermakeup/video/7044215760272493829
[debug] Command-line config: ['-v', '--write-pages', 'https://www.tiktok.com/@mimiermakeup/video/7044215760272493829']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, err UTF-8, pref UTF-8
[debug] yt-dlp version 2022.01.21 [f20d607b0]
[debug] Lazy loading extractors is disabled
[debug] Python version 3.7.3 (CPython 64bit) - Linux-4.9.0-13-amd64-x86_64-with-debian-9.13
[debug] exe versions: ffmpeg 4.2.2-static
[debug] Optional libraries: Cryptodome, mutagen, websockets
[debug] Proxy map: {}
[debug] [TikTok] Extracting URL: https://www.tiktok.com/@mimiermakeup/video/7044215760272493829
[TikTok] 7044215760272493829: Downloading video details
[TikTok] Saving request to 7044215760272493829_https_-_api-h2.tiktokv.com_aweme_v1_aweme_detail_aweme_id=7044215760272493829_version_name=20.1.0_version_code=210_build_number=20.1.0_manifest_version_code=210_update_version_code=210_3c08477359bfe3ef8d9fc55de321bb87.dump
WARNING: [TikTok] 7044215760272493829: Failed to parse JSON  (caused by JSONDecodeError('Expecting value: line 1 column 1 (char 0)')); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with the --verbose flag and include its complete output.; Retrying with feed workaround
[TikTok] 7044215760272493829: Downloading video feed
[TikTok] Saving request to 7044215760272493829_https_-_api-h2.tiktokv.com_aweme_v1_feed_aweme_id=7044215760272493829_version_name=20.1.0_version_code=210_build_number=20.1.0_manifest_version_code=210_update_version_code=210_openudi_c9a154edf052f9316d5180a4a38d586b.dump
[debug] Sort order given by extractor: quality, codec, size, br
[debug] Formats sorted by: hasvid, ie_pref, quality, vcodec, acodec, filesize, fs_approx, tbr, vbr, abr, lang, res, fps, hdr:12(7), asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 7044215760272493829: Downloading 1 format(s): h264_540p_1627798-2
[debug] Invoking downloader on "https://v58.tiktokcdn.com/video/tos/useast2a/tos-useast2a-pve-0068/38e6d69a3ec246c3bf547bf8e99a0945/?VExpiration=1643091560&VSignature=olOHmN5y0Mri3C6SoMiZ9w&a=1180&br=3178&bt=1589&cd=0%7C0%7C0%7C0&ch=0&cr=0&cs=0&cv=1&dr=0&ds=6&er=&ft=.NpOnIX17SBWH6BtReciLo&l=20220125001649010245155110007A61A8&lr=all&mime_type=video_mp4&net=0&pl=0&qs=0&rc=M3E4MzY6ZjM0OjMzNzczM0ApZWhpOTk3PDw6N2hmO2VlZWdpZnIzcjRnNi5gLS1kMTZzc2JeMTNiNWBeMy4uMGBiMTY6Yw%3D%3D&vl=&vr="
[download] Destination: Reply to @its.jazzy.lol3  5 minute rant like always 🤌🏻 [7044215760272493829].mp4
[download] 100% of 28.97MiB in 00:00
```

### PR DESCRIPTION
**Pull request by MinePlayersPE**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
Should handle cases where random versions randomly get blocked

### CODE REVIEW

**Comment by MinePlayersPE** [comment]
they blocked all versions listed here

**Comment by RealDeviil** [comment]
As per https://github.com/yt-dlp/yt-dlp/issues/2476 :

Where are the app and app manifest version numbers coming from?

If I knew where they were coming from, then I would be tempted to automatically update the version numbers (as needed) with whatever the current, working versions are (as per the current TikTok download request).

That approach has to be better than manually hard-coding all known version numbers and then randomly selecting one and/or iterating through them until one works.

I tried finding the version numbers with the Facebook crawler UA, but no joy. Maybe a mobile UA would reveal these numbers?

**Comment by RealDeviil** [comment]
Call me crazy, but is this not the app version here?:  

Screenshot_1

Also, it seems the app manifest version numbers are a direct extension of the app numbers themselves, e.g.:

`_APP_VERSIONS = [('20.9.3', '293'), ('20.4.3', '243'), ('20.2.1', '221'), ('20.1.0', '210'), ('20.0.0', '200')]`

So, I wonder if 22.9.5 and 2295 would work? When I have some more time, I will try it.

In general, if the TikTok app version is published on any viable, reliable/credible site, then we could potentially scrape the version number to dynamically generate the correct version numbers as needed.

**Comment by MinePlayersPE** [comment]
> As per #2476 :
> 
> Where are the app and app manifest version numbers coming from? Are you manually pulling them from the mobile app and/or app URLs?
> 
> If I knew where they were coming from, then I would be tempted to automatically update the version numbers (as needed) with whatever the current, working versions are (as per the current TikTok download request).
> 
> That approach has to be better than manually hard-coding all known version numbers and then randomly selecting one and/or iterating through them until one works.
> 
> I tried finding the version numbers with the Facebook crawler UA (accessing app URLs like https://vm.tiktok.com/ZSe4FqkKd), but no joy. Maybe a mobile UA would reveal these numbers?

They came from... nowhere? @llacb47 initially found 2.0.8/208 version combos by reading through the app data, but iirc the rest are complete guesses

**Comment by RealDeviil** [comment]
> So, I wonder if 22.9.5 and 2295 would work? When I have some more time, I will try it.

So, I tested this, and this theory officially has no basis in reality. :/ I.e., it doesn't work.

**Comment by RealDeviil** [comment]
It also could be that the version numbers are not the real issue.

I can send any version numbers to the "feed" endpoint, and it always works.

**Comment by MinePlayersPE** [comment]
`feed` doesn't even need to take in *any* parameters

**Comment by RealDeviil** [comment]
> They came from... nowhere? @llacb47 initially found 2.0.8/208 version combos by reading through the app data, but iirc the rest are complete guesses

Sounds like we need to know more about **_how_** "@llacb47 initially found 2.0.8/208 version combos by reading through the app data"...That knowledge could be a game changer.

It sure beats (manually) generating (and hard-coding) random numbers and then iterating through them, only to have them all blocked a few hours or days later.

**Comment by MinePlayersPE** [comment]
@RealDeviil he MITM'd the app's network traffic and read the query parameters

**Comment by RealDeviil** [comment]
https://httptoolkit.tech/blog/inspect-any-android-apps-http/

That's the next step to finding a better solution. When that environment is up and running, then inspect the app's HTTP requests/responses to determine the one that reveals TikTok API version numbers. Then see if it's possible to reproduce that request/response outside of the app. Easy enough to try if you already have the environment installed, configured, and ready to go. I personally do not, but if the "feed" endpoint stops working, then maybe I try it.

File: yt_dlp/extractor/tiktok.py [hunk 1]
```diff
@@ -73,16 +85,41 @@ def _call_api(self, ep, query, video_id, fatal=True,
...
-            fatal=fatal, note=note, errnote=errnote, headers={
-                'User-Agent': f'com.ss.android.ugc.trill/{self._MANIFEST_APP_VERSION} (Linux; U; Android 10; en_US; Pixel 4; Build/QQ3A.200805.001; Cronet/58.0.2991.0)',
-                'Accept': 'application/json',
-            }, query=real_query)
+
+    def _call_api(self, ep, query, video_id, fatal=True,
+                  note='Downloading API JSON', errnote='Unable to download API page'):
+        if not self._WORKING_APP_VERSION:
+            app_version = self._configuration_arg('app_version', [''], ie_key=TikTokIE.ie_key())[0]
+            manifest_app_version = self._configuration_arg('manifest_app_version', [''], ie_key=TikTokIE.ie_key())[0]
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:93]: I would have gone with a single args like `tiktok:version=20.9.3,293`. But this is fine too. Either way, document it in readme

File: yt_dlp/extractor/tiktok.py [hunk 2]
```diff
@@ -73,16 +85,41 @@ def _call_api(self, ep, query, video_id, fatal=True,
...
+            elif app_version or manifest_app_version:
+                self.report_warning('Only one of the two required version params are passed as extractor arguments', only_once=True)
+
+        if self._WORKING_APP_VERSION:
+            app_version, manifest_app_version = self._WORKING_APP_VERSION
+            real_query = self._build_api_query(query, app_version, manifest_app_version)
+            return self._call_api_impl(ep, real_query, manifest_app_version, video_id, fatal, note, errnote)
+
+        for count, val in enumerate(self._APP_VERSIONS, start=1):
+            app_version, manifest_app_version = val
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:106]:
```suggestion
        for count, (app_version, manifest_app_version) in enumerate(self._APP_VERSIONS, start=1):
```

**Review by pukkandan — APPROVED** [review]

**Pull request merged**